controller-server/src/main/java/com/yahoo/vespa/hosted/controller/maintenance/BillingReportMaintainer.java

```diff
@@ -73,18 +71,34 @@ List<InvoiceUpdate> maintainInvoices() {
         var tenants = cloudTenants();
         var billsNeedingMaintenance = databaseClient.readBills().stream()
                 .filter(bill -> bill.getExportedId().isPresent())
-                .filter(exported -> exported.status() == BillStatus.OPEN)
+                .filter(exported -> ! exported.status().isFinal())
                 .toList();
 
         for (var bill : billsNeedingMaintenance) {
             var exportedId = bill.getExportedId().orElseThrow();
             var update = reporter.maintainInvoice(tenants.get(bill.tenant()), bill);
-            if (update instanceof ModifiableInvoiceUpdate modifiable && ! modifiable.isEmpty()) {
-                log.fine(invoiceMessage(bill.id(), exportedId) + " was updated with " + modifiable.itemsUpdate());
-            } else if (update instanceof FailedInvoiceUpdate failed && failed.reason == FailedInvoiceUpdate.Reason.REMOVED) {
-                log.fine(invoiceMessage(bill.id(), exportedId) +  " has been deleted in the external system");
-                // Reset the exportedId to null, so that we don't maintain it again
-                databaseClient.setExportedInvoiceId(bill.id(), null);
+            switch (update.type()) {
+                case UNMODIFIED -> log.finer(() ->invoiceMessage(bill.id(), exportedId) + " was not modified");
+                case MODIFIED -> log.fine(invoiceMessage(bill.id(), exportedId) + " was updated with " + update.itemsUpdate().get());
+                case UNMODIFIABLE -> {
+                    if (bill.status() != BillStatus.FROZEN) {
+                        log.fine(() -> invoiceMessage(bill.id(), exportedId) + " is now unmodifiable");
+                        databaseClient.setStatus(bill.id(), "system", BillStatus.FROZEN);
+                    }
+                }
+                case REMOVED -> {
+                    log.fine(() -> invoiceMessage(bill.id(), exportedId) + " has been deleted in the external system");
+                    // Reset the exportedId to null, so that we don't maintain it again
+                    databaseClient.setExportedInvoiceId(bill.id(), null);
+                }
+                case PAID -> {
+                    log.fine(() -> invoiceMessage(bill.id(), exportedId) + " has been paid in the external system");
+                    databaseClient.setStatus(bill.id(), "system", BillStatus.SUCCESSFUL);
+                }
+                case VOIDED -> {
+                    log.fine(() -> invoiceMessage(bill.id(), exportedId) + " has been voided in the external system");
+                    databaseClient.setStatus(bill.id(), "system", BillStatus.VOID);
+                }
```

Comment:
This shows the mapping from invoice update (system-agnostic) to our bill status.